libbfddp/bfddp_extra.c

```diff
@@ -0,0 +1,858 @@
+	 * timer.
+	 */
+	if (bs->bs_state != pstate)
+		bfddp_send_session_state_change(bs);
+
+	/* Notify application about state change. */
+	bfddp_callbacks.bc_state_change(bs, arg, pstate, bs->bs_state);
+}
+
+static void
+bfddp_session_sm_admindown(__attribute__((unused)) struct bfd_session *bs,
+			   __attribute__((unused)) void *arg,
+			   __attribute__((unused)) enum bfd_state_value nstate)
+{
+	/* NOTHING. */
+}
+
+static void
+bfddp_session_sm_down(struct bfd_session *bs, void *arg,
+		      enum bfd_state_value nstate)
+{
+	switch (nstate) {
+	case STATE_ADMINDOWN:
+		/* NOTHING. */
+		break;
+	case STATE_DOWN:
+		bs->bs_state = STATE_INIT;
+
+		/* Notify state change. */
+		bfddp_send_session_state_change(bs);
+
+		/*
+		 * If passive mode we must manually send packet (TX timer is
+		 * disabled).
+		 */
+		if (bs->bs_passive)
+			bfddp_send_control_packet(bs, arg);
+		break;
+	case STATE_INIT:
+		bs->bs_state = STATE_UP;
+
+		/* Start polling. */
+		bs->bs_poll = true;
+
+		/* Notify state change. */
+		bfddp_send_session_state_change(bs);
+		break;
+	case STATE_UP:
+		/* NOTHING: we haven't and the peer hasn't sent INIT yet. */
+		break;
+	}
+}
+
+static void
+bfddp_session_sm_init(struct bfd_session *bs, void *arg,
+		      enum bfd_state_value nstate)
+{
+	switch (nstate) {
+	case STATE_ADMINDOWN:
+		bs->bs_state = STATE_DOWN;
+
+		/* Notify state change. */
+		bfddp_send_session_state_change(bs);
+		break;
+	case STATE_DOWN:
+		/*
+		 * We are waiting peer's INIT.
+		 *
+		 * If we are passive, then we must manually respond to this
+		 * packet (TX timer is disabled).
+		 */
+		if (bs->bs_passive)
+			bfddp_send_control_packet(bs, arg);
+		break;
+	case STATE_INIT:
+		/* FALLTHROUGH. */
+	case STATE_UP:
+		bs->bs_state = STATE_UP;
+		bs->bs_diag = 0;
+
+		/* Start polling. */
+		bs->bs_poll = true;
+
+		/* Notify state change. */
+		bfddp_send_session_state_change(bs);
+		break;
+	}
+}
+
+static void
+bfddp_session_sm_up(struct bfd_session *bs, __attribute__((unused)) void *arg,
+		    enum bfd_state_value nstate)
+{
+	switch (nstate) {
+	case STATE_ADMINDOWN:
+		/* FALLTHROUGH. */
+	case STATE_DOWN:
+		bs->bs_state = STATE_DOWN;
+
+		bfddp_session_set_slowstart(bs);
+		/* Disable echo timers. */
+
+		/* Notify state change. */
+		bfddp_send_session_state_change(bs);
+		break;
+	case STATE_INIT:
+		/* FALLTHROUGH. */
+	case STATE_UP:
+		/* NOTHING. */
+		break;
+	}
+}
+
+void
+bfddp_session_state_machine(struct bfd_session *bs, void *arg,
+			    enum bfd_state_value nstate)
+{
+	enum bfd_state_value ostate = bs->bs_state;
+
+	switch (ostate) {
+	case STATE_ADMINDOWN:
+		bfddp_session_sm_admindown(bs, arg, nstate);
+		break;
+	case STATE_DOWN:
+		bfddp_session_sm_down(bs, arg, nstate);
+		break;
+	case STATE_INIT:
+		bfddp_session_sm_init(bs, arg, nstate);
+		break;
+	case STATE_UP:
+		bfddp_session_sm_up(bs, arg, nstate);
+		break;
+	}
+
+	/* Call application callback to notify about state change. */
+	if (ostate != bs->bs_state) {
```

Comment:
Consider immediate return, to have less indentation:

```
	if (ostate == bs->bs_state) 
              return;

```

Reply:
Makes sense, I'll fix it.